### DISCUSSION TITLE
privateGPT in Docker

### DISCUSSION THREAD

**Discussion by rwcitek** [imartinez/privateGPT#613] [Show and tell]
Greetings!

I've forked the repo and created a branch that puts the instructions in a Dockerfile.  Here's a link to the docker folder in the docker branch in the repo:

https://github.com/rwcitek/privateGPT/tree/docker/docker

I pushed the image that I built to Dockerhub.  So you can get started quickly by simply running this command:

```bash
docker run --rm -it rwcitek/privategpt:2023-06-04 python3 privateGPT.py
```

Regards,
- Robert

**Comment by mvamsee2010** [comment]
Hi. Thank you for adding privategpt to a container image. when I am trying to run this in the terminal I am getting the below error. Would you know what it might be? 
# python3 privateGPT.py
Using embedded DuckDB with persistence: data will be stored in: db
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
Segmentation fault
#

**Comment by rwcitek** [comment]
@mvamsee2010 You're welcome.  And this was just a quick-and-dirty pass, just to see if I could get it running.  Check the issues and pull requests for others who have done more elaborate Dockerfile or docker-compose files.

The "magic" is that you need to be on a system with sufficient RAM, at least 8 GB but 16GB or greater results in a more pleasant experience.  I started on a 2GB system and the image built, but the container instance running privateGPT.py promptly crashed.  I then moved to 16GB, which was much more pleasant.  I haven't tried anything in between.

FYI, since I did my development using [GitHub CodeSpaces]( https://github.com/features/codespaces ), I modified my `devcontainer.json` file to specify 16GB.

https://github.com/rwcitek/privateGPT/blob/docker/.devcontainer/devcontainer.json

Post here letting us know if that does or does not resolve your issue.

**Comment by GoodTimesImmort** [comment]
First of all, thank you for setting this up. After a few tries and an actual distro re-install, I managed to get the docker container running. 
What I would like to do is not have it run the initial digest. I would prefer to load other text to test it out. I am just starting to familiarize myself with docker, so I was wondering if I can do that when I pull the image instead of running it? I'll keep trying to find out on my own, but I thought I'd ask here.

**Reply by rwcitek** [reply]
tl;dr : yes, other text can be loaded.

Details:
1. run `docker run -d --name gpt rwcitek/privategpt sleep inf`  which will start a Docker container instance named `gpt`
2. run `docker container exec gpt rm -rf db/ source_documents/` to remove the existing db/ and source_documents/ folder from the instance
4. run `docker container cp new_text/ gpt:/src/source_documents/` to copy text into the instance from the `new_text/` folder on the host that contains the alternate text you want to use.
5. run `docker container exec gpt python3 ingest.py` to rebuild the db folder, using the new text.
6. run `docker container exec -it gpt python3 privateGPT.py` to run privateGPT with the new text.

And like most things, this is just one of many ways to do it.
I tested the above in a GitHub CodeSpace and it worked.  But post here letting us know how it worked for you.

**Reply by GoodTimesImmort** [reply]
I'm going to give it a try this way. So far I've managed to get the docs I want and even get the ingest.py to run, but the model can't seem to access or reference anything from the new texts, only the state of the union. What I did was as follows. 
1) docker run --rm -it --name gpt rwcitek/privategpt:2023-06-04 python3 privateGPT.py which pulls and runs the container so I end up at the "Enter a query:" prompt (the first ingest has already happened)
2) docker exec -it gpt bash to get shell access
3) rm db and rm source_documents then load text with docker cp 
4) python3 ingest.py in the docker shell
5) Ask questions in the original terminal with the query

**Reply by GoodTimesImmort** [reply]
Test completed and passed with your instructions. I'm not sure what makes the difference yet, but I'm getting results based on the ingested docs I added. Again, thank you for your work and help.

**Comment by dgloeppky** [comment]
Trying to create docker image on a Window 10 PC.

docker build -f Dockerfile.local ,

Gets most of the way through, but towards the end:

 > [dependencies 3/4] RUN poetry install --with local:
4.975 Creating virtualenv private-gpt in /home/worker/app/.venv
5.577 Installing dependencies from lock file
8.424
8.424 Package operations: 148 installs, 1 update, 0 removals
8.424
8.425   • Installing frozenlist (1.4.1)
8.425   • Installing idna (3.6)
8.426   • Installing multidict (6.0.4)
13.60   • Installing aiosignal (1.3.1)
13.60   • Installing attrs (23.1.0)
13.60   • Installing certifi (2023.11.17)
13.61   • Installing charset-normalizer (3.3.2)
13.61   • Installing nvidia-nvjitlink-cu12 (12.3.101)
13.61   • Installing six (1.16.0)
13.61   • Installing urllib3 (1.26.18)
13.62   • Installing wrapt (1.16.0)
13.62   • Installing yarl (1.9.4)
16.00   • Installing aiohttp (3.9.1)
16.00   • Installing dill (0.3.7)
16.00   • Installing filelock (3.13.1)
16.00   • Installing markupsafe (2.1.3)
16.01   • Installing mpmath (1.3.0)
16.01   • Installing fsspec (2023.12.2)
16.01   • Installing deprecated (1.2.14)
16.01   • Installing nvidia-cusparse-cu12 (12.1.0.106)
16.02   • Installing python-dateutil (2.8.2)
16.02   • Installing numpy (1.26.0)
16.02   • Installing nvidia-cublas-cu12 (12.1.3.1)
16.03   • Installing packaging (23.2)
16.18 Connection pool is full, discarding connection: pypi.org. Connection pool size: 10
16.19 Connection pool is full, discarding connection: pypi.org. Connection pool size: 10
18.32   • Installing pytz (2023.3.post1)
18.64   • Installing pyyaml (6.0.1)
18.79   • Installing requests (2.31.0)
18.84   • Installing tqdm (4.66.1)
18.87   • Installing typing-extensions (4.9.0)
18.89   • Installing tzdata (2023.3)
58.38   • Installing h11 (0.14.0)
58.38   • Installing huggingface-hub (0.19.4)
58.38   • Installing humanfriendly (10.0)
58.38   • Installing jinja2 (3.1.2)
58.38   • Installing multiprocess (0.70.15)
58.39   • Installing networkx (3.2.1)
58.39   • Installing nvidia-cuda-cupti-cu12 (12.1.105)
58.39   • Installing nvidia-cudnn-cu12 (8.9.2.26)
58.40   • Installing nvidia-cuda-nvrtc-cu12 (12.1.105)
58.40   • Installing nvidia-cufft-cu12 (11.0.2.54)
58.40   • Installing nvidia-curand-cu12 (10.3.2.106)
58.40   • Installing nvidia-cuda-runtime-cu12 (12.1.105)
58.64 Connection pool is full, discarding connection: pypi.org. Connection pool size: 10
58.83 Connection pool is full, discarding connection: pypi.org. Connection pool size: 10
70.06   • Installing nvidia-cusolver-cu12 (11.4.5.107)
70.08   • Installing nvidia-nccl-cu12 (2.18.1)
70.09   • Installing nvidia-nvtx-cu12 (12.1.105)
70.09   • Installing pandas (2.1.4)
70.11   • Installing protobuf (4.25.1)
70.44   • Installing pyarrow (14.0.1)
71.26   • Updating setuptools (67.6.1 -> 69.0.2)
72.14   • Installing sniffio (1.3.0)
73.13   • Installing sympy (1.12)
77.72   • Installing triton (2.1.0)
78.32   • Installing xxhash (3.4.1)
134.2
134.2   RuntimeError
134.2
134.2   Hash for nvidia-cudnn-cu12 (8.9.2.26) from archive nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl not found in known hashes (was: sha256:b5103041d6e0b8db59159ee5e36b719ba655e63360d490e986c165fa0e63536f)
134.2
134.2   at ~/.local/share/pipx/venvs/poetry/lib/python3.11/site-packages/poetry/installation/executor.py:673 in _validate_archive_hash
134.2       669│         archive_hash: str = "sha256:" + get_file_hash(archive)
134.2       670│         known_hashes = {f["hash"] for f in package.files}
134.2       671│
134.2       672│         if archive_hash not in known_hashes:
134.2     → 673│             raise RuntimeError(
134.2       674│                 f"Hash for {package} from archive {archive.name} not found in"
134.2       675│                 f" known hashes (was: {archive_hash})"
134.2       676│             )
134.2       677│
134.2

Dockerfile.local:26

  24 |     COPY pyproject.toml poetry.lock ./
  25 |
  26 | >>> RUN poetry install --with local
  27 |     RUN poetry install --with ui
  28 |

ERROR: failed to solve: process "/bin/sh -c poetry install --with local" did not complete successfully: exit code: 1

**Comment by Mailaender** [comment]
It would be nice to see an official Docker image.

**Reply by BraintoByte** [reply]
There is a docker-compose.yml file with a docker file that allows you to build it locally on docker

https://github.com/zylon-ai/private-gpt/blob/main/docker-compose.yaml

From the official privateGPT repo
